### DISCUSSION TITLE
Can we invoke runMain on "@main" class?

### DISCUSSION THREAD

**Discussion by hmf** [com-lihaoyi/mill#2663] [Q&A] [answered]
I have the following code snippet:

```scala
object Foo:

  @main
  def main(text: String) =
    println("Hello from @main Foo")

  // def main(args: Array[String]): Unit = 
  //   println("Hello from static Foo main")
```
Here is the error I get:

```shell
hmf@gandalf:/mnt/ssd2/hmf/VSCodeProjects/storch$ ./mill examples.runMain Foo
[114/114] examples.runMain 
Error: Main method not found in class Foo, please define the main method as:
   public static void main(String[] args)
or a JavaFX application class must extend javafx.application.Application
1 targets failed
```
If I use the above comment the "standard" main method, then no error occurs. 

Is their anything I can do, save adding the main method, that lets me execute the "@main"? Reason I ask is because I am creating a Mill build script for an existing SBT based project and I am trying to leave the source as it is. 

TIA

**Comment by lefou** [comment, marked as answer]
I think it should be accessible as just `main` (or maybe `Foo.main`). The [documentation](https://docs.scala-lang.org/scala3/book/methods-main-methods.html) is only 99.9% clear (as there could be multiple object containing a `@main`-annotated `main` in which case it collides).

> A `@main` annotated method can be written either at the top-level (as shown), or inside a statically accessible object. In either case, the name of the program is in each case the name of the method, without any object prefixes.

So, it's either this:

```
$ mill examples.runMain main
```

or maybe this:

```
$ mill examples.runMain Foo.main
```

**Reply by hmf** [reply]
It seems like it is just `main`:

```shell
$ ./mill examples.runMain main "Hello"
[114/114] examples.runMain 
Hello from @main Foo
```
You are also correct about the collision. We need to use another name for the annotated method. 

Thank you.
